libmpv/stream_cb.h

```
@@ -0,0 +1,164 @@
+ * the first argument to all the remaining stream callbacks.
+ */
+
+/**
+ * Opaque context, returned by mpv_get_sub_api(MPV_SUB_API_STREAM_CB).
+ *
+ * This context is global to libmpv, and the callbacks set within it are used
+ * by all mpv_handles created in the same process.
+ */
+typedef struct mpv_stream_cb_context mpv_stream_cb_context;
+
+/**
+ * Open callback used to implement a custom stream.
+ *
+ * @param user_data opaque user data provided via mpv_stream_cb_init()
+ * @param uri name of the stream to be opened
+ * @return opaque cookie identifing the newly opened stream
+ * @return NULL if the URI cannot be opened.
+ */
+typedef void *(*mpv_stream_cb_open_fn)(void *user_data, char *uri);
+
+/**
+ * Read callback used to implement a custom stream.
+ * The semantics of the callback match read(2).
+ *
+ * @param cookie opaque cookie identifying the stream,
+ *               returned from mpv_stream_cb_open_fn
+ * @param buf buffer to read data into
+ * @param size of the buffer
+ * @return number of bytes read into the buffer
+ * @return 0 on EOF
+ * @return -1 on error
+ */
+typedef int64_t (*mpv_stream_cb_read_fn)(void *cookie, char *buf, uint64_t nbytes);
+
+/**
+ * Seek callback used to implement a custom stream.
+ * The semantics of this callback match lseek(2).
+ *
+ * @param cookie opaque cookie identifying the stream,
+ *               returned from mpv_stream_cb_open_fn
+ * @param offset number of bytes to increment or decrement from
+                 the current stream position
+ * @param whence one of SEEK_SET, SEEK_CUR or SEEK_END
+ * @return the resulting offset of the stream
+ * @return -1 if the seek is not allowed or failed
+ */
+typedef int64_t (*mpv_stream_cb_seek_fn)(void *cookie, int64_t offset, int whence);
+
+/**
+ * Size callback used to implement a custom stream.
+ *
+ * @param cookie opaque cookie identifying the stream,
+ *               returned from mpv_stream_cb_open_fn
+ * @return the total size in bytes of the stream
```

IMO should be allowed to fail (which would mean unknown size).